torch/_dynamo/polyfills/functools.py

```diff
@@ -45,3 +45,46 @@ def reduce(
         value = function(value, element)
 
     return value
+
+
+@substitute_class(functools.partial, supports_reconstruction=True)
+class partial:
```

Comment:
Can't we use closure to implement class-like objects?

```python
@substitute_in_graph(functools.partial, is_embedded_type=True)
def partial(func, /, *args, **keywords):
    def newfunc(*fargs, **fkeywords):
        newkeywords = {**keywords, **fkeywords}
        return func(*args, *fargs, **newkeywords)
    newfunc.func = func
    newfunc.args = args
    newfunc.keywords = keywords
    return newfunc
```

Reply:
In the case of `partial`, it could be ok. But, I am thinking in terms of more complicated C classes that have many methods. We can still use closure and add method names as attributes to the returned function object. But, I think its more intuitive for the user to have one-to-one mapping for their C++ class to Python class.

Another discussion point (which is not a class vs closure discussion) is how to reconstruct the value. For example, in your above example, if we want to reconstruct the `partial` object, it will be of type function (and not `functools.partial`). This causes silent correctness issues because the following code could depend on the returned obj being of type `functools.partial`. This is not a class vs closure issue though.

Reply:
The C++ pytree functions can return a C++ object `PyTreeSpec`. The pytree polyfills return a Python object with analog methods while setting `can_constant_fold_through=False`.

See also:

https://github.com/pytorch/pytorch/blob/91c4bf39d39f0607833b2a226e48a9ab7262c906/torch/_dynamo/polyfills/pytree.py#L119-L121

https://github.com/pytorch/pytorch/blob/91c4bf39d39f0607833b2a226e48a9ab7262c906/torch/_dynamo/polyfills/pytree.py#L322-L334

Reply:
For `functools.partial`:

```python
class partial:  # copy the pure-Python version from stdlib
    ...


@@substitute_in_graph(
    functools.partial.__new__, 
    # We need to disable constant folding here because we want the function to reference the 
    # `partial` class defined above, not the one in the C++ module. 
    can_constant_fold_through=False, 
) 
def partial_new(*args, **kwargs):
    return partial(*args, **kwargs)
```

Reply:
These are really good pointers. I think it makes sense to just polyfill `functools.partial.__new__` and provide a helper to convert the already created functools.partial objects to polyfill'd partial objects. 

Let me think more about the reconstruction part.